contrib/pkg/awsprivatelink/disable.go

```diff
@@ -0,0 +1,119 @@
+package awsprivatelink
+
+import (
+	"context"
+	"fmt"
+
+	hiveutils "github.com/openshift/hive/contrib/pkg/utils"
+	awsutils "github.com/openshift/hive/contrib/pkg/utils/aws"
+	hiveclient "github.com/openshift/hive/pkg/client/clientset/versioned"
+	"github.com/openshift/hive/pkg/constants"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+type disableOptions struct {
+	kubernetesClient *kubernetes.Clientset
+	hiveClient       *hiveclient.Clientset
+}
+
+func NewDisableAWSPrivateLinkCommand() *cobra.Command {
+	opt := &disableOptions{}
+
+	cmd := &cobra.Command{
+		Use:   "disable",
+		Short: "Disable AWS PrivateLink",
+		Long: `Disable AWS PrivateLink:
+1) Remove the AWS Hub account credentials Secret created when calling "hiveutil awsprivatelink enable ..."
+2) Empty HiveConfig.spec.awsPrivateLink`,
+		Run: func(cmd *cobra.Command, args []string) {
+			if err := opt.Complete(cmd, args); err != nil {
+				return
+			}
+			if err := opt.Validate(cmd, args); err != nil {
+				return
+			}
+			if err := opt.Run(cmd, args); err != nil {
+				return
+			}
+		},
+	}
+	return cmd
+}
+
+func (o *disableOptions) Complete(cmd *cobra.Command, args []string) error {
+	// Get client config
+	cfg, err := hiveutils.GetClientConfig()
+	if err != nil {
+		log.WithError(err).Fatal("Failed to generate client config")
+	}
+
+	// Get Kubernetes client
+	kubernetesClient, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		log.WithError(err).Fatal("Failed to create Kubernetes clientset")
+	}
+	o.kubernetesClient = kubernetesClient
+
+	// Get Hive client
+	hiveClient, err := hiveclient.NewForConfig(cfg)
+	if err != nil {
+		log.WithError(err).Fatal("Failed to create Hive clientset")
+	}
+	o.hiveClient = hiveClient
+
+	return nil
+}
+
+func (o *disableOptions) Validate(cmd *cobra.Command, args []string) error {
+	return nil
+}
+
+func (o *disableOptions) Run(cmd *cobra.Command, args []string) error {
+	// Get HiveConfig
+	hiveConfig, err := o.hiveClient.HiveV1().HiveConfigs().Get(context.Background(), "hive", metav1.GetOptions{})
+	if err != nil {
+		log.WithError(err).Fatal("Failed to get HiveConfig")
+	}
+	if hiveConfig.Spec.AWSPrivateLink == nil {
+		log.Fatal("AWS PrivateLink is already disabled in HiveConfig")
+	}
+	if len(hiveConfig.Spec.AWSPrivateLink.EndpointVPCInventory) > 0 {
+		log.Fatal("Please remove endpoint VPCs (and the underlying cloud resources) first")
+	}
+
+	// Delete Hub account secret(s) if present
+	hiveNS := hiveConfig.Spec.TargetNamespace
+	if hiveNS == "" {
+		hiveNS = constants.DefaultHiveNamespace
+	}
+	hubAcctSecrets, err := o.kubernetesClient.CoreV1().Secrets(hiveNS).List(
+		context.Background(),
+		metav1.ListOptions{
+			LabelSelector: fmt.Sprintf("%s=true", awsutils.PrivateLinkHubAcctCredLabel),
```

Comment:
So as written, this command will only clean up Secrets created via `enable`. This is as opposed to explicitly looking for the Secret named by HiveConfig.spec.awsPrivateLink.credentialsSecretRef and deleting it. Was that an intentional decision?

Reply:
Yep. I was thinking of the following:
In extreme situations (e.g. "hiveutil awsprivatelink enable ..." fails midway for a few times then reruns successfully), there could be multiple such Secrets, and it would be good to delete all of them.

Reply:
Not sure if the above is a good idea.

Reply:
I've made some changes to follow your suggestion below, i.e. to generate a secret with fixed name each time, proceed if this secret already exists. Only secrets generated with the enable subcommand has the privatelink label, and only the ones with this label are deleted when calling the disable subcommand.